src/Features/CSharp/Portable/Completion/CompletionProviders/InternalsVisibleToCompletionProvider.cs

```diff
@@ -0,0 +1,145 @@
+            if (syntaxTree.IsEntirelyWithinStringLiteral(context.Position, cancellationToken))
+            {
+                var token = syntaxTree.FindTokenOnLeftOfPosition(context.Position, cancellationToken);
+                var attr = GetAttributeSyntaxOfToken(token);
+                if (attr == null)
+                {
+                    return;
+                }
+                if (attr.Name.TryGetNameParts(out IList<string> nameParts) && nameParts.Count > 0)
+                {
+                    var lastName = nameParts[nameParts.Count - 1];
+                    if (lastName == "InternalsVisibleTo" || lastName == "InternalsVisibleToAttribute")
+                    {
+                        AddAssemblyCompletionItems(context, cancellationToken);
+                    }
+                }
+            }
+        }
+
+        private static AttributeSyntax GetAttributeSyntaxOfToken(SyntaxToken token)
+        {
+            //Supported cases:
+            //[Attribute("|
+            //[Attribute(parameterName:"Text|")
+            //Also supported but excluded by syntaxTree.IsEntirelyWithinStringLiteral in ProvideCompletionsAsync
+            //[Attribute(""|
+            //[Attribute("Text"|)
+            var node = token.Parent;
+            if (node is LiteralExpressionSyntax && node.Kind() == SyntaxKind.StringLiteralExpression)
+            {
+                node = node.Parent;
+                if (node is AttributeArgumentSyntax)
+                {
+                    node = node.Parent;
+                    if (node is AttributeArgumentListSyntax)
+                    {
+                        return node.Parent as AttributeSyntax;
+                    }
+                }
+            }
+            return default(AttributeSyntax);
+        }
+
+        private static void AddAssemblyCompletionItems(CompletionContext context, CancellationToken cancellationToken)
+        {
+            var currentProject = context.Document.Project;
+            foreach (var p in context.Document.Project.Solution.Projects)
+            {
+                if (p == currentProject)
+                {
+                    continue;
+                }
+                var completionItem = CompletionItem.Create(displayText: p.AssemblyName,
+                    properties: ImmutableDictionary.Create<string, string>().Add(ProjectIdKey, p.Id.Id.ToString()));
+                context.AddItem(completionItem);
+            }
+        }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = default(char?), CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (item.Properties.TryGetValue(ProjectIdKey, out var projectId))
+            {
+                var assemblyName = item.DisplayText;
+                var project = document.Project.Solution.GetProject(ProjectId.CreateFromSerialized(new System.Guid(projectId)));
```

Comment:
Thats what I had first. I need to get the public key and therefore I need to call GetCompilationAsync on each project. That's why I delayed the calculation to *GetChangeAsync*. I see three options here:

1. Fork the projects as you suggested in an other comment and call GetCompilationAsync in AddAssemblyCompletionItems (in the ProvideCompletionsAsync stage and therefore for every project of the solution).
2. Fork the selected project and call GetCompilationAsync in *GetChangeAsync*. This way I would like to keep the logic to store the project GUID in the properties dictionary for the lookup in *GetChangeAsync*.
3. Bypass *GetCompilationAsync* entirely by calling [StrongNameKeys.Create(options.StrongNameProvider, keyFile, keyContainer, MessageProvider.Instance).PublicKey](https://github.com/dotnet/roslyn/blob/65cc61578e9646cf76a297d8a9e0005afa57378a/src/Compilers/Core/Portable/StrongName/StrongNameKeys.cs#L152). This is not possible at the moment because the *StrongNameKeys* class is not accessible.

I would prefer  option 2. What do you think?